pkg/model/trace/matches.go

```diff
@@ -14,31 +14,23 @@ import (
 
 const RootSpanNotYetReceivedText = "<root span not yet received>"
 
-const (
-	attributesNotFound    = 1
-	attributesFailedMatch = 2
-	attributesPassedMatch = 3
-)
-
 func MatchesProto(id []byte, trace *tempopb.Trace, req *tempopb.SearchRequest) (*tempopb.TraceSearchMetadata, error) {
 	traceStart := uint64(math.MaxUint64)
 	traceEnd := uint64(0)
 
-	traceMatch := attributesNotFound
-	if len(req.Tags) == 0 {
-		traceMatch = attributesPassedMatch
+	// copy the tags from tempopb.SearchRequest. This map is then mutated by the two match* functions below
+	// if, at the end of this loop, there are no more entries in the map then we matched all tags
+	tagsToFind := map[string]string{}
+	for k, v := range req.Tags {
+		tagsToFind[k] = v
 	}
 
 	var rootSpan *v1.Span
 	var rootBatch *v1.ResourceSpans
 	// todo: is it possible to shortcircuit this loop?
 	for _, b := range trace.Batches {
-		// check if the batch matches. we will use the value of batchMatch later to determine if
-		//  - we need to bother checking if the spans match
-		//  - we can mark the entire trace as matching
-		batchMatch := attributesNotFound
-		if traceMatch != attributesPassedMatch {
-			batchMatch = searchAttributes(req.Tags, b.Resource.Attributes)
+		if !allTagsFound(tagsToFind) {
+			matchAttributes(tagsToFind, b.Resource.Attributes)
```

Comment:
Probably good to add a nil check around `b.Resource`.  I think it mostly crops up in testing.